### ISSUE TITLE
Allow to override range in `applyReleasePlan`?

### ISSUE THREAD

**Issue by dotansimha** [atlassian/changesets#426]
## Affected Packages

`@changesets/apply-release-plan`

## Problem

Npm/Yarn doesn't play well with versions like `^1.0.0-alpha.123456` because the caret will apply and look for new version. 
While publishing alpha versions, we would like to make sure that modified `package.json` uses the siblings versions as-is, instead of using it with range modifiers.

In most of our projects, we automate pre-release publishing automatically as part of our CI (we are using a custom script, could be found here: https://github.com/dotansimha/graphql-code-generator/blob/release-pipeline/scripts/canary-release.js#L30-L50 ). We wrote a custom script that uses `changesets` packages to manage the custom versioning. Our goal to is to to look for changed packages, apply new versioning and calculate the new `alpha` version for that package. It means that a change in `minor` will result in bumping minor, and then create a prerelease with the git commit hash.

## Proposed solution

Change this:
https://github.com/atlassian/changesets/blob/master/packages/apply-release-plan/src/version-package.ts#L70

to be:

```ts
          let newNewRange = ignoreRange ? version : `${rangeType}${version}`;
```

And allow passing `ignoreRange` to `applyReleasePlan` (and then pass it to `versionPackage`). 

I tried it with a `patch-package` and it works perfectly and matches our needs. You can find the patch with the changes here: https://github.com/dotansimha/graphql-code-generator/blob/release-pipeline/patches/%40changesets%2Bapply-release-plan%2B4.0.0.patch

**Comment by Andarist** [comment]
I'm wondering if it wouldn't be better to ignore ranges for prereleases altogether (would need to evaluate how to handle peer deps though) as it's possible to retrieve the original range from `initialVersions` when exiting the premode (one extra consideration - how to handle packages created when being in the premode). I agree that prereleases with `^` etc are rather confusing and might lead to obscure issues for people.

This would be a breaking change though, so for the time being we would have to make this an opt-in with a `pre.json` flag or something.

I'm not sure where this logic should be handled though, the distinction between what `@changesets/assemble-release-plan` and `@changesets/apply-release-plan` do in regards to versioning is still a little bit fuzzy to me.

Thoughts? cc @mitchellhamilton

**Comment by mitchellhamilton** [comment]
>I'm wondering if it wouldn't be better to ignore ranges for prereleases altogether (would need to evaluate how to handle peer deps though) as it's possible to retrieve the original range from initialVersions when exiting the premode (one extra consideration - how to handle packages created when being in the premode). I agree that prereleases with ^ etc are rather confusing and might lead to obscure issues for people.

Yes, absolutely. Let's make it an experimental option. Same for snapshot releases.

**Comment by Andarist** [comment]
@dotansimha would you be willing to work on that? 😃

**Comment by dotansimha** [comment]
I think I can give it a try, it depends on how to do you think it should look like. 

If the solution I suggested fits your needs, I can quickly create a PR for it now, if you think it should be implemented in a different way- can you please elaborate?

**Comment by Andarist** [comment]
1. Support for an experimental flag like `useExactDependencyVersioning` should be added to `pre.json` akin flags like this in `config.json`: https://github.com/atlassian/changesets/blob/9dcc364bf19e48f8f2824ebaf967d9ef41b6fc04/packages/config/src/index.test.ts#L31_L34
2. This information should be handled by `@changesets/apply-release-plan` (some other things like `updateInternalDependencies` are handled there)
3. Unfortunately, I was wrong about original ranges being available in `initialVersions` - it's just a flat list of packages after all. Not sure how to best handle this when exiting the pre mode, maybe another flag, something like `defaultDependencyRangeType`, has to be added? Bookkeeping the whole initial dependency tree seems excessive. I would not worry about this one here just yet though - we can figure this out after the initial draft for points 1 and 2 gets in place.

**Comment by mitchellhamilton** [comment]
> Support for an experimental flag like useExactDependencyVersioning should be added to pre.json akin flags like this in config.json

It should be in config.json, not pre.json. We shouldn't encourage people to edit the pre.json.

**Comment by Andarist** [comment]
Fine with me - we should name it `useExactDependencyVersioningInPreMode` then.
